File: ledger-core/src/main/java/com/tradepulse/ledgercore/service/PermissionService.java
package com.tradepulse.ledgercore.service;

import java.util.List;

import org.springframework.stereotype.Service;

import com.tradepulse.ledgercore.exception.ForbiddenException;
import com.tradepulse.ledgercore.repository.RolePermissionRepository;

/**
 * Java-side mirror of the Postgres authorize() function (V6): does any
 * role this user holds (the "user_role" JWT claim, injected by the claim
 * hook in V4/V5) have the requested permission in role_permissions? Same
 * table, same semantics, no implicit admin bypass — just evaluated here
 * because ledger-core's own JDBC connection runs as postgres and never
 * goes through PostgREST/RLS, so Postgres's authorize() is never called
 * on ledger-core's own writes (see V7/V9 comments on why FORCE ROW LEVEL
 * SECURITY is deliberately absent).
 */
@Service
public class PermissionService {

    private final RolePermissionRepository rolePermissionRepository;

    public PermissionService(RolePermissionRepository rolePermissionRepository) {
        this.rolePermissionRepository = rolePermissionRepository;
    }

    public boolean hasPermission(List<String> roles, String permission) {
        if (roles == null || roles.isEmpty()) {
            return false;
        }
        return rolePermissionRepository.existsByPermissionAndRoleIn(permission, roles);
    }

    public void requirePermission(List<String> roles, String permission) {
        if (!hasPermission(roles, permission)) {
            throw ForbiddenException.missingPermission(permission);
        }
    }
}

File: risk-engine/app/permission_service.py
"""Phase 17: Python mirror of ledger-core's PermissionService - same
role_permissions table, same "does any role this user holds have this
permission" check (no implicit admin bypass), evaluated here because
risk-engine enforces its own route-level authorization rather than
delegating to ledger-core. See PermissionService.java's own javadoc for
why this is a plain query rather than going through Postgres's
authorize() function: this connection isn't subject to RLS either.
"""
from typing import List

from fastapi import HTTPException
from sqlalchemy import text
from sqlalchemy.orm import Session


def has_permission(session: Session, roles: List[str], permission: str) -> bool:
    if not roles:
        return False
    result = session.execute(
        text(
            """
            SELECT EXISTS (
                SELECT 1 FROM role_permissions
                WHERE permission = :permission AND role = ANY(:roles)
            )
            """
        ),
        {"permission": permission, "roles": roles},
    ).scalar()
    return bool(result)


def require_permission(session: Session, roles: List[str], permission: str) -> None:
    if not has_permission(session, roles, permission):
        raise HTTPException(status_code=403, detail=f"Missing permission: {permission}")


def roles_from_user(user: dict) -> List[str]:
    """The 'user_role' JWT claim (same claim ledger-core's
    AdminUserController and gateway's SSE ticket flow read), defaulting
    to an empty list rather than raising for a token that somehow
    carries none - has_permission already treats an empty list as no
    permissions, the correct behavior either way."""
    return user.get("user_role") or []
